File: problem/Maximum_Total_Damage/maximumTotalDamage.py
class Solution:
    def maximumTotalDamage(self, power):
        if not power:
            return 0
        freq = {}
        for p in power:
            freq[p] = freq.get(p, 0) + 1
        keys = sorted(freq)
        n = len(keys)
        dp = [0] * n
        dp[0] = freq[keys[0]] * keys[0]
        for i in range(1, n):
            take = freq[keys[i]] * keys[i]
            x = keys[i] - 3
            lo, hi = 0, i
            while lo < hi:
                mid = (lo + hi) // 2
                if keys[mid] <= x:
                    lo = mid + 1
                else:
                    hi = mid
            prev = lo - 1
            if prev >= 0:
                take += dp[prev]
            dp[i] = dp[i - 1] if dp[i - 1] > take else take
        return dp[-1]

File: problem/Maximum_Total_Damage/maximumTotalDamage.js
/**
 * @param {number[]} power
 * @return {number}
 */
var maximumTotalDamage = function(power) {
    const freq = new Map();
    for (const p of power) freq.set(p, (freq.get(p) || 0) + 1);
    const keys = Array.from(freq.keys()).sort((a, b) => a - b);
    const n = keys.length;
    const dp = Array(n).fill(0);
    dp[0] = freq.get(keys[0]) * keys[0];
    for (let i = 1; i < n; i++) {
        let take = freq.get(keys[i]) * keys[i];
        let l = 0, r = i - 1, ans = -1;
        while (l <= r) {
            const mid = Math.floor((l + r) / 2);
            if (keys[mid] <= keys[i] - 3) {
                ans = mid;
                l = mid + 1;
            } else r = mid - 1;
        }
        if (ans >= 0) take += dp[ans];
        dp[i] = Math.max(dp[i - 1], take);
    }
    return dp[n - 1];
};